libraries/lib-music-information-retrieval/MirDsp.cpp

```diff
@@ -0,0 +1,159 @@
+/*  SPDX-License-Identifier: GPL-2.0-or-later */
+/*!********************************************************************
+
+  Audacity: A Digital Audio Editor
+
+  MirDsp.cpp
+
+  Matthieu Hodgkinson
+
+**********************************************************************/
+#include "MirDsp.h"
+#include "FFT.h"
+#include "IteratorX.h"
+#include "MirAudioReader.h"
+#include "MirTypes.h"
+#include "MirUtils.h"
+#include "StftFrameProvider.h"
+
+#include <cassert>
+#include <cmath>
+#include <numeric>
+
+namespace MIR
+{
+namespace
+{
+float GetNoveltyMeasure(
+   const std::vector<float>& prevPowSpec, const std::vector<float>& powSpec)
+{
+   auto k = 0;
+   return std::accumulate(
+      powSpec.begin(), powSpec.end(), 0., [&](float a, float mag) {
+         // Half-wave-rectified stuff
+         return a + std::max(0.f, mag - prevPowSpec[k++]);
+      });
+}
+
+std::vector<float> GetMovingAverage(const std::vector<float>& x, double hopRate)
+{
+   constexpr auto smoothingWindowDuration = 0.2;
+   // An odd number.
+   const int M = std::round(smoothingWindowDuration * hopRate / 4) * 2 + 1;
+   const auto window = GetNormalizedHann(2 * M + 1);
+   auto n = 0;
+   std::vector<float> movingAverage(x.size());
+   std::transform(x.begin(), x.end(), movingAverage.begin(), [&](float) {
+      const auto m = IotaRange(-M, M + 1);
+      const auto y =
+         std::accumulate(m.begin(), m.end(), 0., [&](double y, int i) {
+            auto k = n + i;
+            if (k < 0)
+               k += x.size();
+            else if (k >= x.size())
+               k -= x.size();
```

Comment:
`const auto size = x.size()` (lift that out of the loop)
`k = (k + n + i + size()) % size()`

I don't know whether that might make this hot inner loop run faster -- measure